sdks/python/apache_beam/runners/direct/evaluation_context.py

```diff
@@ -27,22 +27,60 @@
 from apache_beam.runners.direct.watermark_manager import WatermarkManager
 from apache_beam.runners.direct.executor import TransformExecutor
 from apache_beam.runners.direct.direct_metrics import DirectMetrics
+from apache_beam.transforms.trigger import InMemoryUnmergedState
 from apache_beam.utils import counters
 
 
+
+
+class DirectUnmergedState(InMemoryUnmergedState):
+  def __init__(self):
+    super(DirectUnmergedState, self).__init__()
+    self.new_timers = []
+
+  def set_timer(self, window, name, time_domain, timestamp):
+    print '[!!] set_timer', window, name, time_domain, timestamp
+    super(DirectUnmergedState, self).set_timer(window, name, time_domain, timestamp)
+
+  def clear_timer(self, window, name, time_domain):
+    print '[!!] clear_timer', window, name, time_domain
+    # TODO(ccy): this is not implemented, but is not strictly necessary as it is an optimization
+    super(DirectUnmergedState, self).clear_timer(window, name, time_domain)
+
+
+
+class StepContext(object):
+  def __init__(self, execution_context):
+    self._execution_context = execution_context
+    self._state = None
+
+  def get_state(self):
+    # TODO(ccy): consider using copy on write semantics so that work items can be retried.
```

Comment:
<!--new_thread; commit:341fa9245e5d67cc5fab4244bfeca5ea49c72231; resolved:0-->
```python
# TODO(ccy): consider using copy on write semantics so that work items can b
e retried.
```